File: MileageTracker/src/com/switkows/mileage/EditProfiles.java
package com.switkows.mileage;

import android.annotation.TargetApi;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentActivity;
import android.util.SparseBooleanArray;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListAdapter;
import android.widget.ListView;
//import android.widget.TextView;
import android.widget.Toast;

//FIXME - add confirmation when deleting a profile which contains data
//FIXME - upon confirmation (above), delete data associated with profile
public class EditProfiles extends FragmentActivity {
   private final static int EDIT_TEXT_BOX         = 45;

   private ListView mList;
   private int mListViewId;

   private class Profile {
      private long      mId;
      private String    mName;
      private boolean   mHasItems;
      public Profile(long id, String name) {
         mId        = id;
         mName      = name;
         mHasItems  = false;
      }
      public void setHasItems(boolean hasItems) {
         mHasItems = hasItems;
      }

      public String toString() {
         String result = mName;
         if(mHasItems)
            result += "*";
         return result;
      }
      public long getId() {
         return mId;
      }
      public String getName() {
         return mName;
      }
   }

   @Override
   protected void onCreate(Bundle savedInstanceState) {
      super.onCreate(savedInstanceState);
      setContentView(R.layout.edit_profiles);
      if(Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB)
         mListViewId = android.R.layout.simple_list_item_multiple_choice;
      else
         mListViewId = android.R.layout.simple_list_item_activated_1;

      findViewById(R.id.add_profile_button).setOnClickListener(new View.OnClickListener() {
         public void onClick(View v) {
            DialogFragment fragment = CreateProfileDialogFragment.newInstance();
            fragment.show(getSupportFragmentManager(), "dialog");
         }
      });
   }

   @Override
   protected void onResume() {
      ArrayAdapter<Profile> newAdapter = new ArrayAdapter<Profile>(this, mListViewId);

      mList = (ListView)findViewById(android.R.id.list);
      mList.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
      mList.setAdapter(newAdapter);
      updateList();

      super.onResume();
   }

   public void updateList() {
      Cursor cursor = getContentResolver().query(MileageProvider.CAR_PROFILE_URI, null, null, null, null);
      @SuppressWarnings("unchecked")
      ArrayAdapter<Profile> adapter = (ArrayAdapter<Profile>) mList.getAdapter();
      Profile[] content = new Profile[cursor.getCount()];
      int column    = cursor.getColumnIndex(MileageProvider.PROFILE_NAME);
      int idColumn  = cursor.getColumnIndex("_id");
      Uri u;
//      Toast.makeText(this, "Trying to create new ArrayAdapter", Toast.LENGTH_LONG).show();
      for(int i=0 ; i<content.length ; i++) {
         cursor.moveToPosition(i);
         content[i] = new Profile(cursor.getLong(idColumn),cursor.getString(column));
         u = Uri.withAppendedPath(MileageProvider.CAR_CONTENT_URI,content[i].getName());
         Cursor c = getContentResolver().query(u, new String[] {"_id"}, null, null, null);
         if(c.getCount()>0)
            content[i].setHasItems(true);
         c.close();
      }
      cursor.close();
      adapter.clear();
      for (Profile item : content)
         adapter.add(item);
   }

   protected void deleteSelectedProfiles() {
      SparseBooleanArray sel = mList.getCheckedItemPositions();
      long id;
      @SuppressWarnings("unchecked")
      final ArrayAdapter<Profile> arrayAdapter = (ArrayAdapter<Profile>)mList.getAdapter();
      for(int index=0 ; index<sel.size() ; index++) {
         id = arrayAdapter.getItem(sel.keyAt(sel.indexOfValue(true))).getId();
         getContentResolver().delete(ContentUris.withAppendedId(MileageProvider.CAR_PROFILE_URI,id),null,null);
      }
      updateList();
   }
   
   @TargetApi(9)
   protected void renameSelectedProfile(String newName) {
      @SuppressWarnings("unchecked")
      final ArrayAdapter<Profile> arrayAdapter = (ArrayAdapter<Profile>)mList.getAdapter();
      //save off the old value, so we can update records appropriately
      SparseBooleanArray sel = mList.getCheckedItemPositions();
      Profile profile = arrayAdapter.getItem(sel.keyAt(sel.indexOfValue(true)));
      final long oldId = profile.getId();
      String oldName   = profile.getName();
      //get the new value
      ContentValues values = new ContentValues();
      values.put(MileageProvider.PROFILE_NAME, newName);
      //update profile configs
      getContentResolver().update(ContentUris.withAppendedId(MileageProvider.CAR_PROFILE_URI, oldId), values, null, null);
      //move all data to new profile name!
      getContentResolver().update(MileageProvider.ALL_CONTENT_URI, values, MileageData.ToDBNames[MileageData.CAR]+"=?", new String[] {oldName});
      Toast.makeText(getApplicationContext(), "Successfully re-named " + oldName + " to " + newName, Toast.LENGTH_LONG).show();
      String option = getString(R.string.carSelection);
      String currentCar = PreferenceManager.getDefaultSharedPreferences(getApplicationContext()).getString(option, "Car45");
      if(currentCar.compareTo(oldName)==0)
         PreferenceManager.getDefaultSharedPreferences(getApplicationContext()).edit().putString(option, newName).apply();
      updateList();
   }

   @Override
   public boolean onCreateOptionsMenu(Menu menu) {
      MenuInflater inflater = getMenuInflater();
      inflater.inflate(R.menu.edit_profiles_menu, menu);
      return true;
   }
   @Override
   public boolean onOptionsItemSelected(MenuItem item) {
      DialogFragment fragment = null;
      switch (item.getItemId()) {
         case R.id.add_item:
            fragment = CreateProfileDialogFragment.newInstance();
            break;
         case R.id.delete_entry:
            fragment = DeleteProfileDialogFragment.newInstance(getSelectedString());
            break;
         case R.id.change_name:
             SparseBooleanArray sel = mList.getCheckedItemPositions();
             @SuppressWarnings("unchecked")
            final ArrayAdapter<Profile> arrayAdapter = (ArrayAdapter<Profile>)mList.getAdapter();
             String name = arrayAdapter.getItem(sel.keyAt(sel.indexOfValue(true))).getName();

            fragment = EditProfileDialogFragment.newInstance(name);
            break;
      }
      if(fragment != null) {
         fragment.show(getSupportFragmentManager(), "dialog");
         return true;
      }
      return super.onOptionsItemSelected(item);
   }

   private String getSelectedString() {
      SparseBooleanArray selected = mList.getCheckedItemPositions();
      String str = "";
      @SuppressWarnings("unchecked")
      ArrayAdapter<Profile> arrayAdapter = (ArrayAdapter<Profile>)mList.getAdapter();
      for(int i=0 ; i < selected.size() ; i++) {
         if(selected.valueAt(i)) {
            String name = arrayAdapter.getItem(selected.keyAt(i)).toString();
            str = String.format("%s\n%s", str,name);
         }
      }
      return str;
   }
   
   public static class CreateProfileDialogFragment extends DialogFragment {
      public static CreateProfileDialogFragment newInstance() {
         CreateProfileDialogFragment frag = new CreateProfileDialogFragment();
         return frag;
     }
      @Override
      public Dialog onCreateDialog(Bundle savedInstanceState) {
         final EditProfiles activity = (EditProfiles)getActivity();
         OnClickListener acceptListener = new OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
               EditText box = (EditText)((AlertDialog)dialog).findViewById(EDIT_TEXT_BOX);
               String profileName = box.getText().toString();
               MileageProvider.addProfile(activity, profileName);
               activity.updateList();
            }
         };
         EditText editor = new EditText(activity);
         editor.setId(EDIT_TEXT_BOX);
         return new AlertDialog.Builder(activity)
               .setTitle("Enter new Profile name:")
               .setView(editor)
               .setPositiveButton("Add", acceptListener)
               .setNegativeButton("Cancel", new OnClickListener() {
                  public void onClick(DialogInterface dialog, int which) {
                     dialog.dismiss();
                  }})
               .create();
      }
   }

   public static class DeleteProfileDialogFragment extends DialogFragment {
      public static DeleteProfileDialogFragment newInstance(String selected) {
         DeleteProfileDialogFragment frag = new DeleteProfileDialogFragment();
         Bundle args = new Bundle();
         args.putString("selected", selected);
         frag.setArguments(args);
         return frag;
     }
      @Override
      public Dialog onCreateDialog(Bundle savedInstanceState) {
         final EditProfiles activity = (EditProfiles)getActivity();
         String message = getArguments().getString("selected");
         return new AlertDialog.Builder(activity)
               .setTitle("Are you sure you want to delete the following Profiles:")
               .setMessage(message)
               .setPositiveButton("Delete", new OnClickListener() {
                  public void onClick(DialogInterface dialog, int which) {
                     activity.deleteSelectedProfiles();
                  }})
               .setNegativeButton("Cancel", new OnClickListener() {
                  public void onClick(DialogInterface dialog, int which) {
                     dialog.dismiss();
                  }})
               .create();
      }
   }

   public static class EditProfileDialogFragment extends DialogFragment {
      public static EditProfileDialogFragment newInstance(String selected) {
         EditProfileDialogFragment frag = new EditProfileDialogFragment();
         Bundle args = new Bundle();
         args.putString("selected", selected);
         frag.setArguments(args);
         return frag;
     }
      @Override
      public Dialog onCreateDialog(Bundle savedInstanceState) {
         final EditProfiles activity = (EditProfiles)getActivity();
         String name = getArguments().getString("selected");
         if(name.length()==0) {
            Toast.makeText(activity, "Please select a Profile", Toast.LENGTH_SHORT).show();
            return null;
         }
         EditText editor = new EditText(activity);
         editor.setId(EDIT_TEXT_BOX);
         editor.setText(name);
         return new AlertDialog.Builder(activity)
               .setTitle("Enter modified Profile name:")
               .setView(editor)
               .setPositiveButton("Confirm",  new OnClickListener() {
                  public void onClick(DialogInterface dialog, int which) {
                     EditText box = (EditText)((AlertDialog)dialog).findViewById(EDIT_TEXT_BOX);
                     String profileName = box.getText().toString();
                     activity.renameSelectedProfile(profileName);
                  }})
               .setNegativeButton("Cancel", new OnClickListener() {
                  public void onClick(DialogInterface dialog, int which) {
                     dialog.dismiss();
                  }})
               .create();
      }
   }
}

File: app/src/main/java/com/switkows/mileage/EditProfiles.kt
package com.switkows.mileage

import android.annotation.TargetApi
import android.app.AlertDialog
import android.app.Dialog
import android.content.ContentUris
import android.content.ContentValues
import android.content.DialogInterface.OnClickListener
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.ListView
import android.widget.Toast

//FIXME - add confirmation when deleting a profile which contains data
//FIXME - upon confirmation (above), delete data associated with profile
class EditProfiles : FragmentActivity() {
    private var mList: ListView? = null
    private var mListViewId: Int = 0

    private inner class Profile internal constructor(val id: Long, val name: String) {
        private var mHasItems: Boolean = false

        init {
            mHasItems = false
        }

        internal fun setHasItems(hasItems: Boolean) {
            mHasItems = hasItems
        }

        override fun toString(): String {
            var result = name
            if (mHasItems)
                result += "*"
            return result
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.edit_profiles)
        mListViewId = android.R.layout.simple_list_item_activated_1

        findViewById<View>(R.id.add_profile_button).setOnClickListener {
            val fragment = CreateProfileDialogFragment.newInstance()
            fragment.show(supportFragmentManager, "dialog")
        }
    }

    override fun onResume() {
        val newAdapter = ArrayAdapter<Profile>(this, mListViewId)

        mList = findViewById<View>(android.R.id.list) as ListView
        mList!!.choiceMode = ListView.CHOICE_MODE_MULTIPLE
        mList!!.adapter = newAdapter
        updateList()

        super.onResume()
    }

    private fun updateList() {
        val cursor = contentResolver.query(MileageProvider.CAR_PROFILE_URI, null, null, null, null)
        val adapter = mList!!.adapter as ArrayAdapter<Profile>
        if (cursor == null)
            return
        val content = arrayOfNulls<Profile>(cursor.count)
        val column = cursor.getColumnIndex(MileageProvider.PROFILE_NAME)
        val idColumn = cursor.getColumnIndex("_id")
        var u: Uri
        //      Toast.makeText(this, "Trying to create new ArrayAdapter", Toast.LENGTH_LONG).show();
        for (i in content.indices) {
            cursor.moveToPosition(i)
            content[i] = Profile(cursor.getLong(idColumn), cursor.getString(column))
            u = Uri.withAppendedPath(MileageProvider.CAR_CONTENT_URI, content[i]?.name)
            val c = contentResolver.query(u, arrayOf("_id"), null, null, null) ?: continue
            if (c.count > 0)
                content[i]?.setHasItems(true)
            c.close()
        }
        cursor.close()
        adapter.clear()
        content.forEach { item -> adapter.add(item) }
    }

    private fun deleteSelectedProfiles() {
        val sel = mList!!.checkedItemPositions
        var id: Long
        val arrayAdapter = mList!!.adapter as ArrayAdapter<Profile>
        for (index in 0..sel.size() - 1) {
            id = arrayAdapter.getItem(sel.keyAt(sel.indexOfValue(true)))!!.id
            contentResolver.delete(ContentUris.withAppendedId(MileageProvider.CAR_PROFILE_URI, id), null, null)
        }
        updateList()
    }

    @TargetApi(Build.VERSION_CODES.GINGERBREAD)
    private fun renameSelectedProfile(newName: String) {
        val arrayAdapter = mList!!.adapter as ArrayAdapter<Profile>
        //save off the old value, so we can update records appropriately
        val sel = mList!!.checkedItemPositions
        val profile = arrayAdapter.getItem(sel.keyAt(sel.indexOfValue(true)))
        val oldId = profile!!.id
        val oldName = profile.name
        //get the new value
        val values = ContentValues()
        values.put(MileageProvider.PROFILE_NAME, newName)
        //update profile configs
        contentResolver.update(ContentUris.withAppendedId(MileageProvider.CAR_PROFILE_URI, oldId), values, null, null)
        //move all data to new profile name!
        contentResolver.update(MileageProvider.ALL_CONTENT_URI, values, MileageData.ToDBNames[MileageData.CAR] + "=?", arrayOf(oldName))
        Toast.makeText(applicationContext, "Successfully re-named $oldName to $newName", Toast.LENGTH_LONG).show()
        val option = getString(R.string.carSelection)
        val currentCar = PreferenceManager.getDefaultSharedPreferences(applicationContext).getString(option, "Car45")
        if (currentCar!!.compareTo(oldName) == 0)
            PreferenceManager.getDefaultSharedPreferences(applicationContext).edit().putString(option, newName).apply()
        updateList()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.edit_profiles_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        var fragment: DialogFragment? = null
        when (item.itemId) {
            R.id.add_item -> fragment = CreateProfileDialogFragment.newInstance()
            R.id.delete_entry -> fragment = DeleteProfileDialogFragment.newInstance(selectedString)
            R.id.change_name -> {
                val sel = mList!!.checkedItemPositions
                val arrayAdapter = mList!!.adapter as ArrayAdapter<Profile>
                val selected = sel.indexOfValue(true)
                val name = if (selected >= 0) arrayAdapter.getItem(sel.keyAt(selected))?.name else null

                fragment = EditProfileDialogFragment.newInstance(name!!)
            }
        }
        if (fragment != null) {
            fragment.show(supportFragmentManager, "dialog")
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private val selectedString: String
        get() {
            val selected = mList!!.checkedItemPositions
            var str = ""
            val arrayAdapter = mList!!.adapter as ArrayAdapter<Profile>
            (0..selected.size() - 1)
                    .asSequence()
                    .filter { selected.valueAt(it) }
                    .map { arrayAdapter.getItem(selected.keyAt(it))?.toString() }
                    .forEach { str = String.format("%s\n%s", str, it) }
            return str
        }

    class CreateProfileDialogFragment : DialogFragment() {

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val activity = activity as EditProfiles
            val acceptListener = OnClickListener { dialog, _ ->
                val box = (dialog as AlertDialog).findViewById<View>(R.id.edit_text_box) as EditText
                val profileName = box.text.toString()
                MileageProvider.addProfile(activity, profileName)
                activity.updateList()
            }
            val editor = EditText(activity)
            editor.id = R.id.edit_text_box
            return AlertDialog.Builder(activity).setTitle("Enter new Profile name:").setView(editor)
                    .setPositiveButton("Add", acceptListener)
                    .setNegativeButton("Cancel") { dialog, _ -> dialog.dismiss() }.create()
        }

        companion object {
            fun newInstance(): CreateProfileDialogFragment {
                return CreateProfileDialogFragment()
            }
        }
    }

    class DeleteProfileDialogFragment : DialogFragment() {

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val activity = activity as EditProfiles
            val message = arguments.getString("selected")
            return AlertDialog.Builder(activity).setTitle("Are you sure you want to delete the following Profiles:")
                    .setMessage(message)
                    .setPositiveButton("Delete") { _, _ -> activity.deleteSelectedProfiles() }.setNegativeButton("Cancel") { dialog, _ -> dialog.dismiss() }.create()
        }

        companion object {
            fun newInstance(selected: String): DeleteProfileDialogFragment {
                val frag = DeleteProfileDialogFragment()
                val args = Bundle()
                args.putString("selected", selected)
                frag.arguments = args
                return frag
            }
        }
    }

    class EditProfileDialogFragment : DialogFragment() {

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val activity = activity as EditProfiles
            val name = arguments.getString("selected")
            if (name.isEmpty()) {
                return AlertDialog.Builder(activity).setTitle("Error")
                        .setMessage("Select a Profile first").create()
            }
            val editor = EditText(activity)
            editor.id = R.id.edit_text_box
            editor.setText(name)
            return AlertDialog.Builder(activity).setTitle("Enter modified Profile name:").setView(editor)
                    .setPositiveButton("Confirm") { dialog, _ ->
                        val box = (dialog as AlertDialog).findViewById<View>(R.id.edit_text_box) as EditText
                        val profileName = box.text.toString()
                        activity.renameSelectedProfile(profileName)
                    }.setNegativeButton("Cancel") { dialog, _ -> dialog.dismiss() }.create()
        }

        companion object {
            fun newInstance(selected: String): EditProfileDialogFragment {
                val frag = EditProfileDialogFragment()
                val args = Bundle()
                args.putString("selected", selected)
                frag.arguments = args
                return frag
            }
        }
    }
}
